src/App.tsx

```diff
@@ -1,18 +1,95 @@
 import React from 'react';
-import { Routes, Route } from 'react-router-dom';
+import { Route, Routes, useLocation, useNavigate } from 'react-router-dom';
+import Exercises from './pages/Exercises';
+import Home from './pages/Home';
 import Layout from './pages/Layout';
 import NotFound from './pages/NotFound';
-import Home from './pages/Home';
+import Profile from './pages/Profile';
+import SignIn from './pages/SignIn';
+import SignUp from './pages/SignUp';
+import Unauthorized from './pages/Unauthorized';
+import { AuthenticatedUser } from './store/userSlice';
+import useAuth from './utils/useAuth';
+
+interface RouteConfig {
+  element: React.ReactNode;
+  path: string;
+  authHandler: (auth: AuthenticatedUser | null) => boolean;
+}
+
+function ProtectedRoutes({ routes }: { routes: Array<RouteConfig> }) {
+  /** if current route is in routes, check authorization,
+   * nav to Unauthorized if fails
+   */
+  const navigate = useNavigate();
+  /** Guard */
+  const location = useLocation();
+
+  const { authHandler = () => true } = routes.find(({ path }) => location.pathname.includes(path)) ?? {};
+
+  const user = useAuth();
```

Comment:
A part of the problem *could* be that `user` as injected here may not be the same structure/purpose as `user` was in the tutorial/example you pulled this code from. Not sure, but I suspect you could get this functionality working pretty easily with the protected route code in the demos I provided.